File: src/de/umr/fixcon/itarators/SubIterator_FromStart.java
package de.umr.fixcon.itarators;

import com.google.common.collect.Iterables;
import com.google.common.graph.*;
import de.umr.core.Mutator;
import de.umr.core.utils.FastList;

import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import static java.util.stream.Collectors.toList;

public final class SubIterator_FromStart implements Mutator<Graph<Integer>> {

    int searchTreeCounter = 0;
    private final int startVertex;
    private final int targetSize;
    private final MutableGraph<Integer> originalGraph;

    private final MutableGraph<Integer> subgraph = GraphBuilder.undirected().nodeOrder(ElementOrder.insertion()).build();
    private List<Integer> extension_list = new FastList<>();
    private final Deque<Integer> pointerStack = new LinkedList<>(List.of(0));
    private final Deque<Integer> privateStack = new LinkedList<>();

    public SubIterator_FromStart(MutableGraph<Integer> originalGraph, int startVertex, int targetSize) {
        checkArgument(targetSize > 1, "Only supports search for subgraphs of size greater 2");
        this.originalGraph = checkNotNull(originalGraph);
        this.startVertex = startVertex;
        this.targetSize = targetSize;
        subgraph.addNode(startVertex);
        extension_list.addAll(originalGraph.adjacentNodes(startVertex));
        privateStack.push(extension_list.size());
        mutate();
    }

    int startVertex() {
        return startVertex;
    }

    @Override
    public boolean isValid() {
        return subgraph.nodes().size() == targetSize;
    }

    @Override
    public Graph<Integer> current() {
        return subgraph;
    }

    @Override
    public void mutate() {
        do {
            if (subgraph.nodes().size() == targetSize) {
                delete_subset_head();
            } else {
                if (pointerHead_is_outOfRange()) {  //size of subset is < k for following code:
                    if (!pointerHead_is_pending())
                        delete_subset_head();
                    delete_subset_head();
                    pointerStack.pop();
                } else {                            //pivot is not out of range
                    if (pointerHead_is_pending())
                        add_pivot();
                    else
                        pointerStack.push(pointerStack.peek());
                }
            }

        } while (!isValid() && pointerStack.size() > 0);
    }

    /*for the last element in the subset it is not necessary to generate the extension-list, because the subset
    wont be extended further. Therefore in this case the adjustment of the extension-list is omitted.*/
    private void add_pivot() {
        if (subgraph.nodes().size() != targetSize - 1) {
            List<Integer> new_extension = get_new_extension(pivot_vertex());
            extension_list.addAll(new_extension);
            privateStack.push(new_extension.size());
        }
        originalGraph.adjacentNodes(pivot_vertex()).stream().filter(x -> subgraph.nodes().contains(x))
                .forEach(x -> subgraph.putEdge(pivot_vertex(), x));
        pointerStack.push(pointerStack.pop() + 1);
        searchTreeCounter++;
    }

    /*Because for the last element in the subset the extension-list was not adjusted, it also doesn't need
    to be delete here in this case. This is the case if the size of the subset is targetSize*/
    private void delete_subset_head() {
        if (subgraph.nodes().size() != targetSize)
            extension_list = extension_list.subList(0, extension_list.size() - privateStack.pop());
        subgraph.removeNode(Iterables.getLast(subgraph.nodes()));
    }

    private List<Integer> get_new_extension(int pivot_vertex) {
        return originalGraph.adjacentNodes(pivot_vertex).stream()
                .filter(x -> !extension_list.contains(x) && x != startVertex)
                .collect(toList());
    }

    private boolean pointerHead_is_outOfRange() {
        return pointerStack.getFirst() == extension_list.size();
    }

    private boolean pointerHead_is_pending() {
        return subgraph.nodes().size() == pointerStack.size();
    }

    private int pivot_vertex() {
        return extension_list.get(pointerStack.getFirst());
    }
}

File: src/de/umr/fixcon/itarators/SubIteratorFromStart.kt
package de.umr.fixcon.itarators

import de.umr.core.dataStructures.SegmentedList
import de.umr.core.dataStructures.VertexOrderedGraph
import de.umr.core.vertexCount
import de.umr.fixcon.wrappers.CFCO_Problem
import de.umr.fixcon.wrappers.Solution
import org.jgrapht.Graph
import org.jgrapht.Graphs.addEdgeWithVertices
import org.jgrapht.Graphs.neighborSetOf
import org.jgrapht.graph.DefaultEdge
import java.util.*

class SubIteratorFromStart(private val problem: CFCO_Problem, val startVertex: Int, private val currBestSolution: Solution = Solution()) : GraphIterator<Graph<Int, DefaultEdge>> {
    private val subgraph: VertexOrderedGraph<Int> = VertexOrderedGraph(startVertex)
    private var extension = SegmentedList(neighborSetOf(problem.originalGraph, startVertex))
    private val pointerStack = LinkedList(listOf(0))

    init {
        require(problem.targetSize > 1)
        mutate()
    }

    override val isValid get() = numVerticesMissing() == 0

    override val current get() = subgraph

    override fun mutate() {
        do {
            if (isValid) {
                popLastVertexWithExtension()
            } else {
                /**size of subgraph is smaller than targetSize for following code: */
                if (pointerHeadIsOutOfRange()) {
                    if (!pointerHeadIsUnused())
                        popLastVertexWithExtension()
                    popLastVertexWithExtension()
                    pointerStack.pop()
                } else {
                    /**pivot is not out of range*/
                    if (pointerHeadIsUnused()) {
                        addPointerHeadWithExtension()
                    } else {
                        if (pruneWithVertexAdditionBound()) continue
                        pointerStack.push(pointerStack.peek())  //duplicates head element
                    }
                }
            }
        } while (!isValid && pointerStack.isNotEmpty())
        updateSolution()
    }

    private fun updateSolution() {
        if (isValid) currBestSolution.updateIfBetter(current, currentFunctionalValue())
    }

    private fun pruneWithVertexAdditionBound() =
            (currentFunctionalValue() + problem.function.completeAdditionBound(subgraph, problem.targetSize, problem.parameters) <= currBestSolution.value)
                    .also { if (it) popLastVertexWithExtension() }

    private fun currentFunctionalValue() = problem.function.eval(subgraph, problem.parameters)

    private fun numVerticesMissing() = problem.targetSize - subgraph.vertexCount

    private fun addPointerHeadWithExtension() {
        expandExtension()
        addVertexWithEdges(extension[pointerStack.first])
        pointerStack[0]++
    }

    private fun addVertexWithEdges(vertex: Int) =
            neighborSetOf(problem.originalGraph, vertex).filter { subgraph.containsVertex(it) }
                    .forEach { addEdgeWithVertices(subgraph, extension[pointerStack.first], it) }

    private fun popLastVertexWithExtension() = shrinkExtension().also { subgraph.removeLastVertex() }

    /**for the last element in the subset it is not necessary to adjust the extension-list, because the subset
    wont be extended further. Therefore in this case the adjustment of the extension-list is omitted.*/
    private fun expandExtension() {
        if (numVerticesMissing() != 1)
            extension.addAll(neighborSetOf(problem.originalGraph, extension[pointerStack.first])
                    .filter { it !in extension && startVertex != it })
    }

    /**Because for the last element in the subset the extension-list was not adjusted, it also doesn't need
    to be delete here in this case. This is the case if the size of the subset is targetSize*/
    private fun shrinkExtension() {
        if (numVerticesMissing() > 0) extension.removeLastSegment()
    }

    private fun pointerHeadIsOutOfRange() = pointerStack.first == extension.size

    private fun pointerHeadIsUnused() = subgraph.vertexCount == pointerStack.size
}